signer-service/src/api/services/sep10.service.js

```diff
@@ -0,0 +1,95 @@
+const { Keypair } = require('stellar-sdk');
+const { TransactionBuilder, Networks } = require('stellar-sdk');
+const { fetchTomlValues } = require('../helpers/anchors');
+
+const { TOKEN_CONFIG } = require('../../constants/tokenConfig');
+const { CLIENT_SECRET } = require('../../constants/constants');
+
+const NETWORK_PASSPHRASE = Networks.PUBLIC;
+
+exports.signSep10Challenge = async (challengeXDR, outToken, clientPublicKey) => {
+  const clientDomainStellarKeypair = Keypair.fromSecret(CLIENT_SECRET);
+
+  const { signingKey: anchorSigningKey } = await fetchTomlValues(TOKEN_CONFIG[outToken].tomlFileUrl);
+
+  const transactionSigned = new TransactionBuilder.fromXDR(challengeXDR, NETWORK_PASSPHRASE);
+  if (transactionSigned.source !== anchorSigningKey) {
+    throw new Error(`Invalid source account: ${transactionSigned.source}`);
+  }
+  if (transactionSigned.sequence !== '0') {
+    throw new Error(`Invalid sequence number: ${transactionSigned.sequence}`);
+  }
+
+  // See https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0010.md#success
+  // memo field should be empty as we assume (in this implementation) that we use the ephemeral
+  // to authenticate. But no memo sub account derivation.
+  if (transactionSigned.memo.value === '') {
+    throw new Error('Memo does not match');
+  }
+
+  const { operations } = transactionSigned;
+  // Verify the first manage_data operation
+  const firstOp = operations[0];
+  if (firstOp.type !== 'manageData') {
+    throw new Error('The first operation should be manageData');
+  }
+
+  // Only authorize a session that corresponds with the ephemeral client account
+  if (firstOp.source !== clientPublicKey) {
+    throw new Error('First manageData operation must have the client account as the source');
+  }
+
+  const { anchorExpectedKey: expectedKey, clientDomainEnabled } = TOKEN_CONFIG[outToken];
+  if (firstOp.name !== expectedKey) {
+    throw new Error(`First manageData operation should have key '${expectedKey}'`);
+  }
+  if (!firstOp.value || firstOp.value.length !== 64) {
+    throw new Error('First manageData operation should have a 64-byte random nonce as value');
+  }
+
+  // Flags to check presence of required operations
+  let hasWebAuthDomain = false;
+  let hasClientDomain = false;
+
+  // Verify extra manage_data operations, web_auth and proper client domain.
+  for (let i = 1; i < operations.length; i++) {
+    const op = operations[i];
+
+    if (op.type !== 'manageData') {
+      throw new Error('All operations should be manage_data operations');
+    }
+
+    // Verify web_auth_domain operation
+    if (op.name === 'web_auth_domain') {
+      hasWebAuthDomain = true;
+      if (op.source !== anchorSigningKey) {
```

Comment:
I think this can't work. In the previous (first) manage data operation, the `key` is `<home domain> auth`. Here, the value is `web_auth_domain` and in your check this includes the ` auth` suffix as that's how the value is defined in the token config. But the protocol doesn't mention the ' auth' suffix again here. So I guess we need to make sure it's not included (?)

Reply:
This is actually an example of the operation we check (yes, it's confusing that when decoded, `name` is actually `key`:

```
{
    source: 'GDVHOU4AF2QINLYETV2YFC7YWPRVXN4SKR6SOJZ7LAWODJIZJ7ZPJUER',
    type: 'manageData',
    name: 'web_auth_domain',
    value: <Buffer 61 70 69 2e 61 6e 63 6c 61 70 2e 63 6f 6d>
  },
```

We want to ensure the `source` for that op is the one on the `.toml`. Quite like on the first check above. The `source` for the `<home domain> auth` name/key is [expected](https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0010.md#success) to be the client account, in this case the ephemeral.

Reply:
Of course, we should also check the `value` decoding. This I left because it seemed like too many checks for the prototype. After all, we are not doing so in the UI.

Reply:
Ah I see, what confused me is the naming then. In the `tokenConfig.js` file you define `anchorExpectedKey: 'api.anclap.com auth'` and because you gave this 'auth' item a name with `key` I mixed it up with the signing key (that is an actual public key). 

So let's rename `anchorExpectedKey` to something else? Personally, I think it might even be best if we adhere closely to the terminology of the SEP-10 protocol and just define it as `homeDomain: 'api.anclap.com'` and then add the `' auth'` programmatically once needed. Also, we could get rid of the extra declaration of `tomlFileUrl: 'https://api.anclap.com/.well-known/stellar.toml'` as this can also be derived from the `homeDomain` with `https://${homeDomain}/.well-known/stellar.toml`

Reply:
Mmm okay! I like the idea. Less things to keep track of.

Reply:
Wait actually we cannot get rid of the `tomlFile` since for mykobo the domain is `mykobo.co` yet the url for the toml is `https://circle.anchor.mykobo.co/.well-known/stellar.toml`. I think the anchors have some freedom there.

Reply:
Oh 😅 I think Mykobo are breaking the convention then though as the domain should be `circle.anchor.mykobo.co` if I understand the protocol [here](https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0001.md#specification) correctly. But if we can't be certain that our partners adhere to this, then let's keep it separate. Let's still rename the variables maybe?